Tasks/Npm/npm.ts

```diff
@@ -0,0 +1,273 @@
+            break;
+
+        default:
+            tl.setResult(tl.TaskResult.Failed, tl.loc("UnknownCommand", command));
+            return;
+    }
+
+    if (registryLocation == RegistryLocation.Feed) {
+        npmRegistries.push(NpmRegistry.FromFeedId(feedId));
+    } else if (endpointId) {
+        npmRegistries.push(NpmRegistry.FromServiceEndpoint(endpointId, command !== NpmCommand.Publish));
+    }
+
+    for (let registry of npmRegistries) {
+        if (registry.authOnly === false) {
+            appendToNpmrc(npmrc, `registry=${registry.url}\n`);
+        }
+
+        appendToNpmrc(npmrc, `${registry.auth}\n`);
+    }
+
+    await npm.exec();
+
+    tl.rmRF(npmrc);
+    popNpmrc(projectNpmrc);
+    tl.rmRF(getTempPath());
+}
+
+function getTempPath(): string {
+    let tempNpmrcDir
+        = tl.getVariable('Agent.BuildDirectory')
+        || tl.getVariable('Agent.ReleaseDirectory')
+        || process.cwd();
+    let tempPath = path.join(tempNpmrcDir, 'npm');
+    if (tl.exist(tempPath) === false) {
+        tl.mkdirP(tempPath);
+    }
+
+    return tempPath;
+}
+
+function getTempNpmrcPath(): string {
+    let tempNpmrcDir = getTempPath();
+    let tempUserNpmrcPath: string = path.join(
+        tempNpmrcDir,
+        `${tl.getVariable('Build.BuildId')}.npmrc`);
+    return tempUserNpmrcPath;
+}
+
+function appendToNpmrc(npmrc: string, data: string): void {
+    tl.writeFile(npmrc, data, {
+        flag: "a"
+    } as tl.FsOptions);
+}
+
+function pushNpmrc(npmrc: string): void {
+    if (npmrc) {
+        let tempPath = getTempPath();
+        let npmrcPath = path.join(tempPath, ".npmrc");
+
+        if (tl.exist(npmrc)) {
+            tl.debug(tl.loc("SavingProjectNpmrc"));
+            tl.cp(npmrc, path.join(tempPath, ".npmrc"), "-f");
+        }
+    }
+}
+
+function popNpmrc(npmrc: string): void {
+    if (npmrc) {
+        let tempPath = getTempPath();
+        let npmrcPath = path.join(tempPath, ".npmrc");
+
+        if (tl.exist(npmrcPath)) {
+            tl.debug(tl.loc("RestoringProjectNpmrc"));
+            tl.cp(npmrcPath, npmrc, "-f");
+            tl.rmRF(npmrcPath);
+        }
+    }
+}
+
+function getFeedIdFromRegistry(registry: string) {
+    let registryUrl = url.parse(registry);
+    let registryPathname = registryUrl.pathname.toLowerCase();
+    let startingToken = "/_packaging/";
+    let startingIndex = registryPathname.indexOf(startingToken);
+    let endingIndex = registryPathname.indexOf("/npm/registry");
+
+    return registryUrl.pathname.substring(startingIndex + startingToken.length, endingIndex);
+}
+
+function GetLocalRegistries(npmrc: string): string[] {
+    let localRegistries: string[] = [];
+    let registries = GetRegistries(npmrc);
+    let collectionUrl = url.parse(tl.getVariable("System.TeamFoundationCollectionUri"));
+
+    for (let registry of registries) {
+        let registryUrl = url.parse(registry);
+        if (collectionUrl.host.toLowerCase() === collectionUrl.host.toLowerCase()) {
```

Comment:
You're comparing the URL against itself.
And this probably won't work in hosted since registryUrl.host != collectionUrl.host